replicator/src/main/scala/com/evolutiongaming/kafka/journal/replicator/TopicCommit.scala

```diff
@@ -1,44 +1,132 @@
 import com.evolutiongaming.kafka.journal.KafkaConsumer
+import com.evolutiongaming.kafka.journal.replicator.commit.AsyncPeriodicTopicCommit
 import com.evolutiongaming.kafka.journal.util.TemporalHelper.*
 import com.evolutiongaming.skafka.*
+import com.evolutiongaming.skafka.consumer.RebalanceCallback
 
 import java.time.Instant
+import scala.annotation.nowarn
 import scala.collection.immutable.SortedMap
 import scala.concurrent.duration.*
 
-trait TopicCommit[F[_]] {
+/**
+ * Kafka partition offset commit logic for replicator internal topic processing API
+ * ([[TopicConsumer]], [[ConsumeTopic]]).
+ *
+ * @tparam F effect type
+ */
+private[journal] trait TopicCommit[F[_]] {
+// journal-private because the places where it is used are journal-private - TopicConsumer, ConsumeTopic
 
+  /**
+   * Mark offsets for commit.
+   * 
+   * The method should be called after each processing step, i.e., after each consumer poll result processing.
+   * 
+   * Please note that offsets passed here should follow the same logic as for Java consumer commit methods:
+   * if offset N is known to be processed, offset N+1 should be committed - offset from which you want to start your
+   * processing next time.
+   * 
+   * Depending on the underlying implementation, this method might execute commit synchronously with the call or
+   * schedule it for later.
+   * 
+   * @param offsets partition offsets to commit
+   */
   def apply(offsets: Nem[Partition, Offset]): F[Unit]
+
+  /**
+   * Notify topic commit logic about new partitions assigned during rebalance - called in
+   * [[com.evolutiongaming.skafka.consumer.RebalanceListener1.onPartitionsAssigned]].
+   * 
+   * @param partitions assigned partitions
+   *                   
+   * @return [[RebalanceCallback]]
+   */
+  def onPartitionsAssigned(@nowarn("cat=unused") partitions: NonEmptySet[Partition]): RebalanceCallback[F, Unit] =
+    RebalanceCallback.pure(())
+
+  /**
+   * Notify topic commit logic about partitions revoked during rebalance - called in
+   * [[com.evolutiongaming.skafka.consumer.RebalanceListener1.onPartitionsRevoked]].
+   * 
+   * Could be used to commit offsets for revoked partitions using
+   * [[RebalanceCallback.commit]] in order not to lose progress during rebalance.
+   *
+   * @param partitions revoked partitions
+   *
+   * @return [[RebalanceCallback]]
+   */
+  def onPartitionsRevoked(@nowarn("cat=unused") partitions: NonEmptySet[Partition]): RebalanceCallback[F, Unit] =
+    RebalanceCallback.pure(())
+
+  /**
+   * Notify topic commit logic about partitions lost - called in
+   * [[com.evolutiongaming.skafka.consumer.RebalanceListener1.onPartitionsLost]].
+   *
+   * When partitions are lost as opposed to being revoked:
+   * [[org.apache.kafka.clients.consumer.ConsumerRebalanceListener#onPartitionsLost(java.util.Collection)]]
+   * 
+   * Usually nothing could be done in this case apart from removing the lost partitions from internal state, if there
+   * is any.
```

Comment:
shall we add more letter, like:
```suggestion
   * Usually nothing could be done in this case apart from removing the lost partitions from internal state, if there
   * is any. As lost partitions usually are already handled by other consumer, it is recommended to NOT commit offsets concurrently.
```

Reply:
There is a link to `org.apache.kafka.clients.consumer.ConsumerRebalanceListener#onPartitionsLost(java.util.Collection)` above, I don't it is a good idea to duplicate information from it here.

Reply:
IMHO having clearer docs is beneficial, even if it duplicates line, which are about 10 navigation jumps away in IDE

Reply:
I like to find important information at 1-click distance, not somewhere far and deep away

Reply:
added additional info the doc about concurrent commits